File: 290_wordPattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        s = s.split()
        if len(pattern) != len(s): 
            return False
        hash_map = {}
        for i in range(len(pattern)):
            if pattern[i] not in hash_map:
                if s[i] in hash_map.values():
                    return False
                hash_map[pattern[i]] = s[i]
            else:
                if hash_map[pattern[i]] != s[i]:
                    return False
        return True

print(Solution().wordPattern("aaaa", "dog cat cat dog")) 

File: 290_wordPattern.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:

    vector<string> splitStringIntoWords(string s){
        stringstream ss(s); 
        string word;      
        vector<string> words; 
        while (ss >> word) {
            words.push_back(word);
        }
        return words;        
    }

    bool wordPattern(string pattern, string s) {

        unordered_map<char,string> comp;
        int n= pattern.size();
        vector<string> words = splitStringIntoWords(s);
        for(int i=0; i<n; i++){
            comp[pattern[i]]=words[i];
        }
        int j=0;
        for(auto i:comp){
            if(i.first!=pattern[j]) return false;
            j++;
        }
        return true;

        
    }
};

int main(){
    Solution sol;

    string pattern ="abba",s="dog cat cat dog";
    
    cout<<sol.wordPattern(pattern,s)<<"\n";
    return 0;
}